### PR TITLE
Implement CSI NodeStageVolume, NodeUnstageVolume

### LINKED ISSUE #97: Need to implement NodeStageVolume/NodeUnstageVolume
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind feature

**What happened**: Failed to start CSI plugin pods on my setup.

**What you expected to happen**: After "kubectl create -f opensds/nbp/csi/server/deploy/kubernetes", I expect 3 CSI plugin pods up and running, but they failed to start.

**How to reproduce it (as minimally and precisely as possible)**:
Deployed using opensds-installer ansible in csi mode.  CSI plugin pods failed to start.

xyang@ubuntu:~/go/src/github.com/opensds/nbp/csi/server/deploy$ kubectl get pod
NAME                                 READY     STATUS             RESTARTS   AGE
csi-attacher-opensdsplugin-0         0/2       CrashLoopBackOff   12         6m
csi-nodeplugin-opensdsplugin-9drgf   0/2       CrashLoopBackOff   12         6m
csi-provisioner-opensdsplugin-0      0/2       CrashLoopBackOff   12         6m

---------------------------------------------------------
$ kubectl describe pod csi-provisioner-opensdsplugin-0 
Events:
  Type     Reason                 Age              From                Message
  ----     ------                 ----             ----                -------
  Normal   Scheduled              4m               default-scheduler   Successfully assigned csi-provisioner-opensdsplugin-0 to 127.0.0.1
  Normal   SuccessfulMountVolume  4m               kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "socket-dir"
  Normal   SuccessfulMountVolume  4m               kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "csi-provisioner-token-ds6vq"
  Normal   Created                4m (x3 over 4m)  kubelet, 127.0.0.1  Created container
  Warning  Failed                 4m (x3 over 4m)  kubelet, 127.0.0.1  Error: failed to start container "csi-provisioner": Error response from daemon: linux mounts: Could not find source mount of /var/lib/kubelet/pods/af360ec3-6cbc-11e8-95d2-000c29f57be0/volumes/kubernetes.io~empty-dir/socket-dir
  Normal   Pulled                 4m (x3 over 4m)  kubelet, 127.0.0.1  Container image "opensdsio/csiplugin:latest" already present on machine
  Normal   Created                4m (x3 over 4m)  kubelet, 127.0.0.1  Created container
  Warning  Failed                 4m (x3 over 4m)  kubelet, 127.0.0.1  Error: failed to start container "opensds": Error response from daemon: linux mounts: Could not find source mount of /var/lib/kubelet/pods/af360ec3-6cbc-11e8-95d2-000c29f57be0/volumes/kubernetes.io~empty-dir/socket-dir
  Warning  BackOff                4m (x2 over 4m)  kubelet, 127.0.0.1  Back-off restarting failed container
  Warning  BackOff                4m (x2 over 4m)  kubelet, 127.0.0.1  Back-off restarting failed container
  Normal   Pulled                 4m (x4 over 4m)  kubelet, 127.0.0.1  Container image "quay.io/k8scsi/csi-provisioner:v0.2.0" already present on machine

------------------
kubectl describe pod csi-nodeplugin-opensdsplugin-9drgf

Events:
  Type     Reason                 Age               From                Message
  ----     ------                 ----              ----                -------
  Normal   SuccessfulMountVolume  9m                kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "pods-mount-dir"
  Normal   SuccessfulMountVolume  9m                kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "pods-probe-dir"
  Normal   SuccessfulMountVolume  9m                kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "iscsi-dir"
  Normal   SuccessfulMountVolume  9m                kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "ceph-dir"
  Normal   SuccessfulMountVolume  9m                kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "socket-dir"
  Normal   SuccessfulMountVolume  9m                kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "csi-nodeplugin-token-dnff5"
  Warning  Failed                 9m                kubelet, 127.0.0.1  Error: failed to start container "opensds": Error response from daemon: linux mounts: Could not find source mount of /etc/ceph
  Normal   Created                9m (x3 over 9m)   kubelet, 127.0.0.1  Created container
  Warning  Failed                 9m (x3 over 9m)   kubelet, 127.0.0.1  Error: failed to start container "driver-registrar": Error response from daemon: linux mounts: Could not find source mount of /var/lib/kubelet/plugins/csi-opensdsplugin
  Normal   Pulled                 9m (x3 over 9m)   kubelet, 127.0.0.1  Container image "opensdsio/csiplugin:latest" already present on machine
  Normal   Created                9m (x3 over 9m)   kubelet, 127.0.0.1  Created container
  Normal   Pulled                 9m (x3 over 9m)   kubelet, 127.0.0.1  Container image "quay.io/k8scsi/driver-registrar:v0.2.0" already present on machine
  Warning  Failed                 9m (x2 over 9m)   kubelet, 127.0.0.1  Error: failed to start container "opensds": Error response from daemon: linux mounts: Could not find source mount of /etc/iscsi
  Warning  BackOff                4m (x23 over 9m)  kubelet, 127.0.0.1  Back-off restarting failed container

---------------------------------------------------------
kubectl describe pod csi-attacher-opensdsplugin-0

Events:
  Type     Reason                 Age               From                Message
  ----     ------                 ----              ----                -------
  Normal   Scheduled              9m                default-scheduler   Successfully assigned csi-attacher-opensdsplugin-0 to 127.0.0.1
  Normal   SuccessfulMountVolume  9m                kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "iscsi-dir"
  Normal   SuccessfulMountVolume  9m                kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "ceph-dir"
  Normal   SuccessfulMountVolume  9m                kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "socket-dir"
  Normal   SuccessfulMountVolume  9m                kubelet, 127.0.0.1  MountVolume.SetUp succeeded for volume "csi-attacher-token-vn2kt"
  Normal   Created                9m (x3 over 9m)   kubelet, 127.0.0.1  Created container
  Normal   Pulled                 9m (x3 over 9m)   kubelet, 127.0.0.1  Container image "quay.io/k8scsi/csi-attacher:v0.2.0" already present on machine
  Warning  Failed                 9m (x3 over 9m)   kubelet, 127.0.0.1  Error: failed to start container "csi-attacher": Error response from daemon: linux mounts: Could not find source mount of /var/lib/kubelet/pods/af319f82-6cbc-11e8-95d2-000c29f57be0/volumes/kubernetes.io~empty-dir/socket-dir
  Normal   Pulled                 9m (x3 over 9m)   kubelet, 127.0.0.1  Container image "opensdsio/csiplugin:latest" already present on machine
  Normal   Created                9m (x3 over 9m)   kubelet, 127.0.0.1  Created container
  Warning  Failed                 9m (x3 over 9m)   kubelet, 127.0.0.1  Error: failed to start container "opensds": Error response from daemon: linux mounts: Could not find source mount of /var/lib/kubelet/pods/af319f82-6cbc-11e8-95d2-000c29f57be0/volumes/kubernetes.io~empty-dir/socket-dir
  Warning  BackOff                9m (x2 over 9m)   kubelet, 127.0.0.1  Back-off restarting failed container
  Warning  BackOff                4m (x24 over 9m)  kubelet, 127.0.0.1  Back-off restarting failed container

-----------------------------
But socket-dir is available:

root@ubuntu:/var/lib/kubelet/pods/af360ec3-6cbc-11e8-95d2-000c29f57be0/volumes/kubernetes.io~empty-dir# ls -l
total 4
drwxrwxrwx 2 root root 4096 Jun 10 07:43 socket-dir

After reading this issue https://github.com/kubernetes/kubernetes-anywhere/issues/88, I did a bind mount:
sudo mount -o bind /var/lib/kubelet /var/lib/kubelet
sudo mount --make-shared /var/lib/kubelet

sudo mount -o bind /etc/iscsi /etc/iscsi
sudo mount --make-shared /etc/iscsi

sudo mount -o bind /etc/ceph /etc/ceph
sudo mount --make-shared /etc/ceph

Deleted the pods and recreated them.  After that the provisioner and attacher pod was started successfully:

xyang@ubuntu:~/go/src/github.com/opensds/nbp/csi/server/deploy$ kubectl get pod
NAME                                 READY     STATUS             RESTARTS   AGE
csi-attacher-opensdsplugin-0         2/2       Running            0          4m
csi-nodeplugin-opensdsplugin-sdhs2   1/2       CrashLoopBackOff   5          4m
csi-provisioner-opensdsplugin-0      2/2       Running            0          4m

NodePublishVolume is the place to do "bind mount", while NodeStageVolume is the place to do a mount at a "global" dir.  We should implement NodeStageVolume and do bind mount in NodePublishVolume.

**Anything else we need to know?**:

**Environment**:
- NBP version:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by BaiHuoYu**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #97 
Implement CSI NodeStageVolume, NodeUnstageVolume

**Special notes for your reviewer**:

**Release note**:

```release-note
Implements NodeStageVolume and NodeUnstageVolume for CSI driver.
```

### CODE REVIEW

File: csi/server/plugin/opensds/node_test.go [hunk 1]
```diff
@@ -52,7 +52,7 @@ func TestNodeGetCapabilities(t *testing.T) {
 		&csi.NodeServiceCapability{
 			Type: &csi.NodeServiceCapability_Rpc{
 				Rpc: &csi.NodeServiceCapability_RPC{
-					Type: csi.NodeServiceCapability_RPC_UNKNOWN,
+					Type: csi.NodeServiceCapability_RPC_STAGE_UNSTAGE_VOLUME,
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node_test.go]: Need to add unit tests to do stage/unstage volume, not just report capabilities.

Also there are no tests for node publish/unpublish.  Need to add tests for that too but that can be a separate PR.

Comment (BaiHuoYu) [reply csi/server/plugin/opensds/node_test.go]: Currently these can only be tested manually, but not unit tests.

**Comment by xing-yang** [comment]
There are lots of redundant code between NodeStageVolume and NodePublishVolume, and between NodeUnstageVolume and NodeUnpublishVolume.  Can you move the common code to a util function?

**Comment by xing-yang** [comment]
Also please resolve conflicts.

**Comment by leonwanghui** [comment]
@BaiHuoYu Please resolve these errors

**Comment by xing-yang** [comment]
Please add a release note.

**Comment by xing-yang** [comment]
@BaiHuoYu For release note, you can just write down something like "Implements CSI NodeStageVolume and NodeUnstageVolume.".

**Comment by xing-yang** [comment]
@BaiHuoYu I added release notes at the top.

File: csi/server/plugin/opensds/node.go [hunk 2]
```diff
@@ -17,258 +17,424 @@ package opensds
...
 
 ////////////////////////////////////////////////////////////////////////////////
 //                            Node Service                                    //
 ////////////////////////////////////////////////////////////////////////////////
 
-func (p *Plugin) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+var (
+	// Client opensds client
+	Client *c.Client
```
Comment (xing-yang) [inline csi/server/plugin/opensds/node.go]: I downloaded this PR but can't even build it.

root@ubuntu:~/gopath/src/github.com/opensds/nbp# make
mkdir -p  /root/gopath/src/github.com/opensds/nbp/build/out
go build -o /root/gopath/src/github.com/opensds/nbp/build/out/csi.server.opensds github.com/opensds/nbp/csi/server
# github.com/opensds/nbp/csi/server/plugin/opensds
csi/server/plugin/opensds/node.go:41:2: Client redeclared in this block
	previous declaration at csi/server/plugin/opensds/controller.go:42:2

**Comment by xing-yang** [comment]
Tested this PR with the changes mentioned in my comments.  I'll submit another PR to address them.
